File: server/projects.ts
import { readdirSync, readFileSync, existsSync, writeFileSync } from "node:fs";
import { join, relative, resolve, sep } from "node:path";
import { loadConfig } from "./config.ts";
import { getCurrentBranch, getLastCommitTs, listWorktrees } from "./git.ts";

export interface ProjectSummary {
  name: string;
  path: string;
  currentBranch: string | null;
  detectedCommand: string | null;
  devCommand: string;
  hasEnv: boolean;
  packageManager: string;
  lastCommitTs: number | null;
}

function detectPackageManager(dir: string): string {
  if (existsSync(join(dir, "pnpm-lock.yaml"))) return "pnpm";
  if (existsSync(join(dir, "yarn.lock"))) return "yarn";
  if (existsSync(join(dir, "bun.lockb")) || existsSync(join(dir, "bun.lock"))) return "bun";
  return "npm";
}

function runScript(pm: string, script: string): string {
  if (pm === "npm") return `npm run ${script}`;
  return `${pm} ${script}`; // pnpm dev / yarn dev / bun dev
}

/** Guess the dev command from package.json scripts. */
export function detectDevCommand(dir: string): string | null {
  const pkgPath = join(dir, "package.json");
  if (!existsSync(pkgPath)) return null;
  try {
    const pkg = JSON.parse(readFileSync(pkgPath, "utf8"));
    const scripts: Record<string, string> = pkg.scripts ?? {};
    const pm = detectPackageManager(dir);
    for (const candidate of loadConfig().devScriptPriority) {
      if (scripts[candidate]) return runScript(pm, candidate);
    }
    return null;
  } catch {
    return null;
  }
}

export function resolveDevCommand(dir: string): string {
  const cfg = loadConfig();
  const override = cfg.overrides[dir]?.devCommand;
  if (override && override.trim()) return override.trim();
  return detectDevCommand(dir) ?? "npm run dev";
}

/** Directory names we never descend into while searching for repos. */
const SKIP_DIRS = new Set([
  "node_modules",
  ".git",
  "dist",
  "build",
  ".next",
  ".turbo",
  ".cache",
  "coverage",
  "vendor",
]);

/**
 * Recursively find git repos under `root`, descending at most `maxDepth` levels.
 * Stops at the first `.git` on any branch, so a repo's own subfolders aren't
 * treated as separate projects. Returns absolute repo directories.
 */
function findGitRepos(root: string, maxDepth: number): string[] {
  const found: string[] = [];
  const walk = (dir: string, depth: number) => {
    // A folder containing .git is a repo — record it and don't descend further.
    if (existsSync(join(dir, ".git"))) {
      found.push(dir);
      return;
    }
    if (depth >= maxDepth) return;
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const e of entries) {
      if (!e.isDirectory() && !e.isSymbolicLink()) continue;
      if (e.name.startsWith(".") || SKIP_DIRS.has(e.name)) continue;
      walk(join(dir, e.name), depth + 1);
    }
  };
  let rootEntries;
  try {
    rootEntries = readdirSync(root, { withFileTypes: true });
  } catch {
    return found;
  }
  for (const e of rootEntries) {
    if (!e.isDirectory() && !e.isSymbolicLink()) continue;
    if (e.name.startsWith(".") || SKIP_DIRS.has(e.name)) continue;
    walk(join(root, e.name), 1);
  }
  return found;
}

export async function listProjects(): Promise<ProjectSummary[]> {
  const cfg = loadConfig();
  const parent = cfg.parentDir;
  if (!existsSync(parent)) return [];

  const repoDirs = findGitRepos(parent, Math.max(1, cfg.maxScanDepth));

  // Enrich each repo concurrently — git calls dominate the wall time.
  const scanned = await Promise.all(
    repoDirs.map(async (dir): Promise<ProjectSummary | null> => {
      const hasPackageJson = existsSync(join(dir, "package.json"));
      if (!hasPackageJson && !cfg.showNonNodeProjects) return null;

      const detected = detectDevCommand(dir);
      const pm = detectPackageManager(dir);
      const [currentBranch, lastCommitTs] = await Promise.all([
        getCurrentBranch(dir),
        getLastCommitTs(dir),
      ]);
      // Name is the path relative to the scanning folder, so nested repos
      // (e.g. "sweet/frontend/app") stay grouped and unambiguous.
      const name = relative(parent, dir).split(sep).join("/");
      return {
        name,
        path: dir,
        currentBranch,
        detectedCommand: detected,
        devCommand: resolveDevCommand(dir),
        hasEnv: cfg.envFiles.some((f) => existsSync(join(dir, f))),
        packageManager: pm,
        lastCommitTs,
      };
    }),
  );

  const results = scanned.filter((p): p is ProjectSummary => p !== null);
  // Most recently changed first; repos with no commits sink to the bottom.
  results.sort((a, b) => {
    const ta = a.lastCommitTs ?? -Infinity;
    const tb = b.lastCommitTs ?? -Infinity;
    if (tb !== ta) return tb - ta;
    return a.name.localeCompare(b.name);
  });
  return results;
}

export interface EnvFile {
  name: string;
  exists: boolean;
  content: string;
}

export function readEnvFiles(dir: string): EnvFile[] {
  return loadConfig().envFiles.map((name) => {
    const p = join(dir, name);
    const exists = existsSync(p);
    return { name, exists, content: exists ? readFileSync(p, "utf8") : "" };
  });
}

export function writeEnvFile(dir: string, name: string, content: string): void {
  if (!loadConfig().envFiles.includes(name)) {
    throw new Error(`Refusing to write unknown env file: ${name}`);
  }
  writeFileSync(join(dir, name), content);
}

/** Resolve a project path from its (possibly nested) name, guarding against traversal. */
export function projectPathFromName(name: string): string {
  const cfg = loadConfig();
  const parent = resolve(cfg.parentDir);
  const dir = resolve(parent, name);
  // Must stay inside the scanning folder — blocks "../" escapes.
  if (dir !== parent && !dir.startsWith(parent + sep)) {
    throw new Error("Invalid project name");
  }
  return dir;
}

/**
 * Resolve a working directory for a project: the main repo, or one of its
 * worktrees. Validates cwd against the actual worktree list (worktrees can live
 * anywhere on disk), so arbitrary paths are rejected.
 */
export async function resolveWorkdir(name: string, cwd?: string): Promise<string> {
  const main = projectPathFromName(name);
  if (!cwd || resolve(cwd) === resolve(main)) return main;
  const target = resolve(cwd);
  const worktrees = await listWorktrees(main);
  const allowed = worktrees.some((w) => resolve(w.path) === target);
  if (!allowed) throw new Error("Directory is not a worktree of this project");
  return target;
}

File: src-tauri/src/projects.rs
//! Project scanning, dev-command detection, env files, and path guards —
//! mirrors server/projects.ts.
use crate::config;
use crate::git;
use serde::Serialize;
use std::path::{Component, Path, PathBuf};

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSummary {
    pub name: String,
    pub path: String,
    pub current_branch: Option<String>,
    pub detected_command: Option<String>,
    pub dev_command: String,
    pub has_env: bool,
    pub package_manager: String,
    pub last_commit_ts: Option<i64>,
}

#[derive(Debug, Serialize)]
pub struct EnvFile {
    pub name: String,
    pub exists: bool,
    pub content: String,
}

const SKIP_DIRS: &[&str] = &[
    "node_modules",
    ".git",
    "dist",
    "build",
    ".next",
    ".turbo",
    ".cache",
    "coverage",
    "vendor",
];

fn detect_package_manager(dir: &Path) -> String {
    if dir.join("pnpm-lock.yaml").exists() {
        "pnpm".into()
    } else if dir.join("yarn.lock").exists() {
        "yarn".into()
    } else if dir.join("bun.lockb").exists() || dir.join("bun.lock").exists() {
        "bun".into()
    } else {
        "npm".into()
    }
}

fn run_script(pm: &str, script: &str) -> String {
    if pm == "npm" {
        format!("npm run {script}")
    } else {
        format!("{pm} {script}")
    }
}

pub fn detect_dev_command(dir: &Path) -> Option<String> {
    let pkg_path = dir.join("package.json");
    if !pkg_path.exists() {
        return None;
    }
    let raw = std::fs::read_to_string(&pkg_path).ok()?;
    let pkg: serde_json::Value = serde_json::from_str(&raw).ok()?;
    let scripts = pkg.get("scripts");
    let pm = detect_package_manager(dir);
    for candidate in config::load().dev_script_priority {
        let has = scripts
            .and_then(|s| s.get(&candidate))
            .map(|v| !v.is_null())
            .unwrap_or(false);
        if has {
            return Some(run_script(&pm, &candidate));
        }
    }
    None
}

pub fn resolve_dev_command(dir: &Path) -> String {
    let cfg = config::load();
    if let Some(ov) = cfg.overrides.get(&dir.to_string_lossy().to_string()) {
        if let Some(cmd) = &ov.dev_command {
            if !cmd.trim().is_empty() {
                return cmd.trim().to_string();
            }
        }
    }
    detect_dev_command(dir).unwrap_or_else(|| "npm run dev".to_string())
}

fn walk(dir: &Path, depth: u32, max_depth: u32, found: &mut Vec<PathBuf>) {
    if dir.join(".git").exists() {
        found.push(dir.to_path_buf());
        return;
    }
    if depth >= max_depth {
        return;
    }
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };
    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().to_string();
        if name.starts_with('.') || SKIP_DIRS.contains(&name.as_str()) {
            continue;
        }
        let path = entry.path();
        if !path.is_dir() {
            continue;
        }
        walk(&path, depth + 1, max_depth, found);
    }
}

fn find_git_repos(root: &Path, max_depth: u32) -> Vec<PathBuf> {
    let mut found = Vec::new();
    let entries = match std::fs::read_dir(root) {
        Ok(e) => e,
        Err(_) => return found,
    };
    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().to_string();
        if name.starts_with('.') || SKIP_DIRS.contains(&name.as_str()) {
            continue;
        }
        let path = entry.path();
        if !path.is_dir() {
            continue;
        }
        walk(&path, 1, max_depth, &mut found);
    }
    found
}

pub fn list_projects() -> Vec<ProjectSummary> {
    let cfg = config::load();
    let parent = PathBuf::from(&cfg.parent_dir);
    if !parent.exists() {
        return vec![];
    }
    let repos = find_git_repos(&parent, cfg.max_scan_depth.max(1));
    let mut out: Vec<ProjectSummary> = Vec::new();
    for dir in repos {
        let has_pkg = dir.join("package.json").exists();
        if !has_pkg && !cfg.show_non_node_projects {
            continue;
        }
        let dir_str = dir.to_string_lossy().to_string();
        let detected = detect_dev_command(&dir);
        let pm = detect_package_manager(&dir);
        let current_branch = git::current_branch(&dir_str);
        let last_commit_ts = git::last_commit_ts(&dir_str);
        let name = dir
            .strip_prefix(&parent)
            .map(|p| p.components().map(|c| c.as_os_str().to_string_lossy()).collect::<Vec<_>>().join("/"))
            .unwrap_or_else(|_| dir_str.clone());
        let has_env = cfg.env_files.iter().any(|f| dir.join(f).exists());
        out.push(ProjectSummary {
            name,
            path: dir_str,
            current_branch,
            detected_command: detected.clone(),
            dev_command: resolve_dev_command(&dir),
            has_env,
            package_manager: pm,
            last_commit_ts,
        });
    }
    // Most recently changed first; repos with no commits sink to the bottom; ties by name.
    out.sort_by(|a, b| {
        let ta = a.last_commit_ts.unwrap_or(i64::MIN);
        let tb = b.last_commit_ts.unwrap_or(i64::MIN);
        tb.cmp(&ta).then_with(|| a.name.cmp(&b.name))
    });
    out
}

pub fn read_env_files(dir: &str) -> Vec<EnvFile> {
    config::load()
        .env_files
        .iter()
        .map(|name| {
            let p = Path::new(dir).join(name);
            let exists = p.exists();
            let content = if exists {
                std::fs::read_to_string(&p).unwrap_or_default()
            } else {
                String::new()
            };
            EnvFile {
                name: name.clone(),
                exists,
                content,
            }
        })
        .collect()
}

pub fn write_env_file(dir: &str, name: &str, content: &str) -> Result<(), String> {
    if !config::load().env_files.iter().any(|f| f == name) {
        return Err(format!("Refusing to write unknown env file: {name}"));
    }
    std::fs::write(Path::new(dir).join(name), content).map_err(|e| e.to_string())
}

/// Lexically normalise a path (collapse "." and ".."), like Node's path.normalize.
fn normalize(path: &Path) -> PathBuf {
    let mut out: Vec<Component> = Vec::new();
    for comp in path.components() {
        match comp {
            Component::CurDir => {}
            Component::ParentDir => match out.last() {
                Some(Component::Normal(_)) => {
                    out.pop();
                }
                Some(Component::RootDir) | Some(Component::Prefix(_)) => {}
                _ => out.push(comp),
            },
            c => out.push(c),
        }
    }
    out.iter().collect()
}

/// Absolute, normalised path (relative inputs resolved against the process cwd) —
/// mirrors Node's path.resolve.
fn resolve_abs(p: &str) -> PathBuf {
    let path = Path::new(p);
    let joined = if path.is_absolute() {
        path.to_path_buf()
    } else {
        std::env::current_dir().unwrap_or_default().join(path)
    };
    normalize(&joined)
}

/// Resolve a project path from its (possibly nested) name, guarding against traversal.
pub fn project_path_from_name(name: &str) -> Result<String, String> {
    let cfg = config::load();
    let parent = resolve_abs(&cfg.parent_dir);
    let dir = normalize(&parent.join(name));
    if dir.starts_with(&parent) {
        Ok(dir.to_string_lossy().to_string())
    } else {
        Err("Invalid project name".to_string())
    }
}

/// Resolve a working dir: the main repo or one of its worktrees. Rejects arbitrary paths.
pub fn resolve_workdir(name: &str, cwd: Option<&str>) -> Result<String, String> {
    let main = project_path_from_name(name)?;
    let main_abs = resolve_abs(&main);
    match cwd {
        None => Ok(main),
        Some(c) if c.is_empty() => Ok(main),
        Some(c) => {
            let target = resolve_abs(c);
            if target == main_abs {
                return Ok(main);
            }
            let worktrees = git::list_worktrees(&main);
            if worktrees.iter().any(|w| resolve_abs(&w.path) == target) {
                Ok(target.to_string_lossy().to_string())
            } else {
                Err("Directory is not a worktree of this project".to_string())
            }
        }
    }
}
